processing/data_collection/gazette/middlewares.py

```diff
@@ -1,13 +1,72 @@
-# -*- coding: utf-8 -*-
-# Define here the models for your spider middleware
-#
-# See documentation in:
-# https://doc.scrapy.org/en/latest/topics/spider-middleware.html
-from scrapy import signals
+import datetime
+import logging
 
+from scrapy import Request, signals
 
-class GazetteSpiderMiddleware:
 
+class GazetteDateFilteringSpiderMiddleware:
+    @classmethod
+    def from_crawler(cls, crawler):
+        s = cls()
+        s.logger = logging.getLogger(f"{__name__}.{cls.__name__}")
+        return s
+
+    def process_spider_output(self, response, result, spider):
+        for request_or_item in result:
+            if isinstance(request_or_item, Request):
+                if request_or_item.dont_filter is True:
+                    yield request_or_item
+                    continue
+
+                gazette_date = request_or_item.meta.get("gazette_date")
```

Comment:
This Middleware assumes that we know beforehand what dates the request will return. This is not always true. For example, in a paginated page, one request can return data from from different years/months/days. Also, if we can ensure that a request will return information from a specific date/period, as we have this information and `start_date` available in the spider, we could simply do the filtering in the spider, avoiding the use of a MIddleware that may or may not work in a more general basis.

Reply:
> This Middleware assumes that we know beforehand what dates the request will return. This is not always true. For example, in a paginated page, one request can return data from from different years/months/days.

Correct me if I made a mistake as I'm not sure I understood this particular concern. But this implementation allows not filtering requests for not having the date information and when `dont_filter` in present in the meta. So, we could agree that filtering with the middleware and filtering inside the spider would provide the same functionality?

Also, it allows partial dates like "only having the year" or "only having year and month".

For pagination, using `sp_jundiai` as an example, information about the gazettes dates could be extracted and used in the requests for the gazette info page and also for the pagination if we used the date of the last displayed gazette in the pagination request (since the system shows gazettes in reverse sorted manner, this shouldn't be a problem). Also, we could not give any info about the date in the pagination requests, if that logic is not correct. I think this implementation would be pretty similar to what we would be able to do inside the spider, right? For others systems that don't show this information, we can't do much about that either way :/

> Also, if we can ensure that a request will return information from a specific date/period, as we have this information and start_date available in the spider, we could simply do the filtering in the spider, avoiding the use of a MIddleware that may or may not work in a more general basis.

If the functionality of doing it inside the spiders is the same, why wouldn't it be better to make a middleware responsible for this? Just like the `GazetteDateFilteringPipeline` is responsible for filtering the items instead of writing it in every spider, don't you think we should centralize it?

If we were to add another argument like `end_date` or something we haven't thought about right now, all spiders would have to be refactored again if that code belonged in every spider, right?